packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx

```diff
@@ -276,14 +224,88 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
     if (valueChanged) {
       roundedValue = precisionRound(newValue!, precision);
       setCurrentValue(roundedValue);
+    } else if (displayValueChanged && !isControlled) {
+      const nextValue = parseFloat(newDisplayValue as string);
+      if (!isNaN(nextValue)) {
+        setCurrentValue(precisionRound(nextValue, precision));
+      }
     }
 
     if (valueChanged || displayValueChanged) {
       onChange?.(e, { value: roundedValue, displayValue: newDisplayValue });
     }
+
+    setTextValue(undefined);
   };
 
-  state.input.value = textValue;
+  const state: SpinButtonState = {
+    size,
+    appearance,
+    spinState: keyboardSpinState,
+    atBound: internalState.current.atBound,
+
+    components: {
+      root: 'span',
+      input: 'input',
+      incrementButton: 'button',
+      decrementButton: 'button',
+    },
+    root: resolveShorthand(root, {
+      required: true,
+      defaultProps: nativeProps.root,
+    }),
+    input: resolveShorthand(input, {
+      required: true,
+      defaultProps: {
+        ref,
+        autoComplete: 'off',
+        role: 'spinbutton',
+        appearance: appearance,
+        type: 'text',
+        ...nativeProps.primary,
+      },
+    }),
+    incrementButton: resolveShorthand(incrementButton, {
+      required: true,
+      defaultProps: {
+        tabIndex: -1,
+        children: <ChevronUp16Regular />,
+        disabled: nativeProps.primary.disabled,
+        'aria-label': 'Increment value',
+        type: 'button',
+      },
+    }),
+    decrementButton: resolveShorthand(decrementButton, {
+      required: true,
+      defaultProps: {
+        tabIndex: -1,
+        children: <ChevronDown16Regular />,
+        disabled: nativeProps.primary.disabled,
+        'aria-label': 'Decrement value',
+        type: 'button',
+      },
+    }),
+  };
+
+  let valueToDisplay;
+  if (textValue !== undefined) {
+    valueToDisplay = textValue;
+  } else if (value === null || currentValue === null) {
+    valueToDisplay = displayValue ?? '';
+    internalState.current.value = null;
+    internalState.current.atBound = 'none';
```

Comment:
FYI: writing to `.current` has consequences with Concurrent Mode and can cause issues, https://frontarm.com/daishi-kato/use-ref-in-concurrent-mode/. I don't that current code will cause problems, but it should be verified once `@fluentui/react-components` will support React 18.